## Aim:
How to link the rule-based extracted CI_TYPE-GEO pairs with the respective Ci failure impacts

**Idea**\
Test using prompt engineering by passing table of CIGEO pairs to GPT-J model.
Steps:
* Load model and apply it always on one chunk of the document to extract CI failure impacts 
* Use prompt engineering to extract time and location of the CI failure (origin) the CI impacts (impact location)
or 
* Pass dataframe of pairs as input to the model
or
* Use few shot prompting with example answers

**Finally:**
* Compaire all approaches of spatial and temporal linking CI failure impacts

In [1]:
%env CUDA_DEVICE_ORDER=PCI_BUS_ID
%env CUDA_VISIBLE_DEVICES=0  # nvidia gpu
%env PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
# %env TORCH_CUDA_ARCH_LIST=8.6

# settings for distributed computing
%env WORLD_SIZE=1
%env RANK=0
%env LOCAL_RANK=0

# NOTE: # WORLD_SIZE: each GPU corresponds to one process (world = no. of processes within a group), processes communicate with each other enabling eg., distributed training
# NOTE: # RANK: IDs of the processes, ranging from 0 up to WORLD_SIZE - 1

env: CUDA_DEVICE_ORDER=PCI_BUS_ID
env: CUDA_VISIBLE_DEVICES=0  # nvidia gpu
env: PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
env: WORLD_SIZE=1
env: RANK=0
env: LOCAL_RANK=0


In [2]:
import os
import sys
import re
import glob
from pathlib import Path
from io import StringIO

import numpy as np
import pandas as pd
from jinja2 import Template
from langchain_docling import DoclingLoader
import spacy
from huggingface_hub import login
from transformers import (
    AutoTokenizer,
    AutoModelForCausalLM,
    BitsAndBytesConfig,
    GPTJForQuestionAnswering
)
import torch
import transformers


sys.path.append("../")
import src.settings as s

torch.manual_seed(42)

# set default location to store model before loading transformers
os.environ["HF_HOME"] = (
    "/home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval/notebooks/huggingface_mirror/"
)

/home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval/.venv/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
# %env MASTER_ADDR=127.0.0.1
# %env MASTER_PORT=6006

# # # Initialize distributed computing
# rank = int(os.environ["RANK"])
# device = torch.device(f"cuda:{rank}")
# torch.cuda.set_device(device)
# torch.distributed.init_process_group(backend="nccl")

In [4]:
pd.set_option(
    "display.max_colwidth", None
)  #  automatic linebreaks and multi-line cells.
pd.set_option("display.colheader_justify", "left")

## Generate CI_GEO-pairs

In [5]:
import importlib.util
# import spacy_transformers


try:
    nlp = spacy.load(s.settings.SPACY_MODEL)
except (OSError, ValueError) as e:
    print(f"spaCy language model '{s.settings.SPACY_MODEL}' not found. Downloading ...")
    ## loading transformer language model for NER requires additional package
    if (s.settings.SPACY_MODEL.endswith("_trf") and importlib.util.find_spec("spacy[transformers]") is None):
        !uv add spacy[transformers]
    !uv run python -m spacy download {s.settings.SPACY_MODEL}
    nlp = spacy.load(s.settings.SPACY_MODEL)

print(f"Loaded spaCy language model: {s.settings.SPACY_MODEL}")

Loaded spaCy language model: en_core_web_trf


In [14]:
## Create New entity for transport infrastructure and apply it on any doc

## see for more info: https://spacy.io/usage/rule-based-matching#entityruler
## NOTE EntityRuler is hidden inside .add_pipe()


## call nlp model and create pipeline with new entity pattern
# NOTE Creating new entity (CI_TYPE) solves the issue that FAC entities (buildings, airports, highways, bridges, etc.) refer only to the name of the facility (e.g. A76, Ahrtalbahn)
config = {"spans_key": None, "annotate_ents": True, "overwrite": False}
try:
    ruler = nlp.add_pipe("span_ruler", config=config)
    ruler.from_disk("../ner_patterns.jsonl")
except ValueError:
    print("SpanRuler already exists in pipeline.")
    ruler = nlp.get_pipe("span_ruler")
    ruler.from_disk("../ner_patterns.jsonl")


# store patterns in jsonl file, example:
# ruler.add_patterns([
#     {"label": "CI_TYPE", "pattern": "road?.+"},
#    {"label":"CI_TYPE","pattern":"rail.*$"},
# ])
# ruler.to_disk("../ner_patterns.jsonl")


## load doc
PARSED_TEXT_DIR = "../" + s.settings.PATH_DATA + "parsed_documents/"
FILE_PATH = PARSED_TEXT_DIR + "Koks et al 2022 Brief communication_cleaned.md"
loader = DoclingLoader(FILE_PATH)  # use chunks from Docling.Loader
doc = loader.load()

## TODO make as pydantic class with fixed attributes
df_ci_geo = pd.DataFrame(
    columns=[
        "chunk_id",
        "ci_entity",
        "ci_entity_label",
        "geo_entity",
        "geo_entity_label",
        "token_distance",
    ]
)


## get most likely geolocation for each CI entity based on distance
for i, chunk in enumerate(doc):
    nlp_chunk = nlp(chunk.page_content)
    all_ents = [ent for ent in nlp_chunk.ents]
    ci_type_ents = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE", "FAC"]]
    ci_type_ents_info = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE"]]
    fac_ents_info = [ent for ent in nlp_chunk.ents if ent.label_ in ["FAC"]]

    # check if chunk contains CI_TYPE entities
    if len(ci_type_ents) > 0:
        print(f"\nChunk [{i}], No. CI_TYPE and FAC entities: {len(ci_type_ents)}")
        print(
            f"Contains following entities for CI_TYPE: {ci_type_ents_info}, FAC: {fac_ents_info}"
        )
        print(f"Chunk text [{i}]:", chunk.page_content)
        # print(f"{ {(ci_type_ents[i].text, ci_type_ents[i].label_) for i in range(len(ci_type_ents))} } ")

        # iterate over all entities within chunk
        for ent_idx in range(len(all_ents)):
            # when entity is CI_TYPE or FAC (i.e. buidling, airports, highways) do following ...
            if all_ents[ent_idx].label_ in ["CI_TYPE", "FAC"]:
                ci_idx = ent_idx

                ## .. calculate distances between CI_TYPE entity and  all GEO entities in chunk based on index position
                distance_list = []
                idx_in_chunk = []
                try:
                    for ent_idx in range(len(all_ents)):
                        # TODO calc distances between CI_TYPE ~ GEO entities based on word numbers and not entities
                        if all_ents[ent_idx].label_ in ["GPE", "LOC"]:
                            geo_idx = ent_idx
                            dist_ent_pair = np.abs(ci_idx - geo_idx)
                            distance_list.append(dist_ent_pair)
                            idx_in_chunk.append((ent_idx))
                            closest_pair_idx = np.argmin(
                                distance_list
                            )  # idx of closest GEO entity
                            distance_closest_pair = distance_list[closest_pair_idx]

                    threshold = 5  # max token distance between CI_TYPE and GEO entity
                    if distance_closest_pair > threshold:
                        print(
                            f""" Token distance between CI_TYPE/FAR and next GEO entity is {distance_closest_pair} and thus larger than the allowed distance of {threshold} tokens """
                        )
                        continue
                    else:
                        print(
                            f"""  Closest GEO entity to CI_TYPE/FAC entity "{all_ents[ci_idx]}" is "{all_ents[idx_in_chunk[closest_pair_idx]]}" at distance {distance_closest_pair}"""
                        )  # TODO constrain min.distance to max value (eg. 5 tokens), issue: likely when distance value is high that geolocation of Ci_type is mentioned in previous sentences or chunk

                    ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
                    result_dict = {
                        "chunk_id": i,
                        "ci_entity": all_ents[ci_idx].text,
                        "ci_entity_label": all_ents[ci_idx].label_,
                        "geo_entity": all_ents[idx_in_chunk[closest_pair_idx]].text,
                        "geo_entity_label": all_ents[
                            idx_in_chunk[closest_pair_idx]
                        ].label_,
                        "token_distance": distance_closest_pair,
                    }
                    df_ci_geo = pd.concat(
                        [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
                    )

                except IndexError:
                    print("No GEO entities found in this chunk.")
                    continue
                # print("\nidx_in_chunk, closest pair idx", idx_in_chunk, closest_pair_idx)

                # spacy.displacy.render(
                #     nlp_chunk, style="ent",
                #     options={"ents": ["CI_TYPE", "GPE", "LOC", "FAC"], "colors": {"CI_TYPE": "violet"}}
                # )

        else:
            print("\nNo CI_TYPE or FAC entities found in this chunk.")
            continue

SpanRuler already exists in pipeline.


2025-12-04 23:00:54,839 - INFO - Going to convert document batch...
2025-12-04 23:00:54,840 - INFO - Initializing pipeline for SimplePipeline with options hash 4cc01982ae99b46a2a63fcda46c47c35
2025-12-04 23:00:54,840 - INFO - Processing document Koks et al 2022 Brief communication_cleaned.md
2025-12-04 23:00:54,880 - INFO - Finished converting document Koks et al 2022 Brief communication_cleaned.md in 0.04 sec.
Token indices sequence length is longer than the specified maximum sequence length for this model (556 > 512). Running this sequence through the model will result in indexing errors



Chunk [1], No. CI_TYPE and FAC entities: 3
Contains following entities for CI_TYPE: [bridges, schools, hospitals], FAC: []
Chunk text [1]: Abstract. Germany, Belgium and the Netherlands were hit by extreme precipitation and flooding in July 2021. This brief communication provides an overview of the impacts to large-scale critical infrastructure systems and how recovery has progressed. The results show that Germany and Belgium were particularly affected, with many infrastructure assets severely damaged or completely destroyed. Impacts range from completely destroyed bridges and sewage systems, to severely damaged schools and hospitals. We find that (large-scale) risk assessments, often focused on larger (river) flood events, do not find these local, but severe, impacts due to critical infrastructure failures. This may be the result of limited availability of validation material. As such, this brief communication not only will help to better understand how critical infrastructure can be aff

In [15]:
df_ci_geo.loc[df_ci_geo["ci_entity"] == "rail"]  # .head(15)


,chunk_id,ci_entity,ci_entity_label,geo_entity,geo_entity_label,token_distance
17,6,rail,CI_TYPE,the Ahr valley,LOC,2
46,11,rail,CI_TYPE,Altenburg,GPE,4


In [17]:
doc[13].page_content

'We found no information regarding direct impact on solid-waste facilities as a result of the flood event. However, there is a large pressure on the solid-waste sector to clean the affected areas; 1 month after the event, we observed dozens of large temporary waste fills and frequent incidences of oil pollution in Rhineland-Palatinate during a field visit. In the Ahrweiler district alone, the flood caused as much solid waste as normally would be collected over 30 years. In Belgium, the amount of solid waste is estimated around 160 000 t, stored at several places, such as the abandoned highway track A601. This highway has been used for approximately 9 months as a temporary storage for debris (Couplez, 2022). In the Netherlands, there have been primarily problems with waste deposits along the river banks, which is mostly the solid waste transported by the river from further upstream. Thousands of tonnes of tree debris (logs and\nNat. Hazards Earth Syst. Sci., 22, 3831–3838, 2022\nE. E. Koks 

## Apply GPT-J on chunks // few-shot prompting


In [ ]:
import torch
import gc

print(torch.cuda.memory_summary(device=None, abbreviated=False))
# # empyty CUDA cache
gc.collect()
torch.cuda.empty_cache()
# print(torch.cuda.memory_summary(device=None, abbreviated=False))

|===========================================================================|
|                  PyTorch CUDA memory summary, device ID 0                 |
|---------------------------------------------------------------------------|
|            CUDA OOMs: 0            |        cudaMalloc retries: 0         |
|===========================================================================|
|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |
|---------------------------------------------------------------------------|
| Allocated memory      |      0 B   |      0 B   |      0 B   |      0 B   |
|       from large pool |      0 B   |      0 B   |      0 B   |      0 B   |
|       from small pool |      0 B   |      0 B   |      0 B   |      0 B   |
|---------------------------------------------------------------------------|
| Active memory         |      0 B   |      0 B   |      0 B   |      0 B   |
|       from large pool |      0 B   |      0 B   |      0 B   |

In [ ]:
question = "Which impacts of infrastructure failures are mentioned in the text? Categorize the output by the type of infrastructure, societal or economic impacts, the location and possibly the time of the infrastructure failure."

In [ ]:
## add CITATIONS
# and refer to the citations provided in the context.

#####################
## LLAMA (improved columns for ci_types, locations, damages, economic impacts were still mixed-up)
#     Evaluate and improve your answer based on the information about critical infrastructure (CI) types (column: "ci_entity") and their geolocations (column: "geo_entity") mentioned in CI locations.
#     Question: "{{ question }}"

prompt_template = """

    You are an expert analyst assistant and should use ONLY the provided context to answer the following question:
    
    Question: "{{ question }}"

    The examples below should guide you in creating the answer.

    Then, evaluate and improve the information about infrastructure types (ci_entity) and their geolocation (geo_entity) provided in the columns of the CI locations dataset to improve your answer.


    CI locations:
    {% for item in context %}
    - ("ci_entity" and "geo_entity":\n {{ item["ci_locations"][["ci_entity", "geo_entity"]] }})
    {% endfor %}
    

    Return ONLY valid JSON in the following list format:
    [{
        "infrastructure_type": "...",
        "damage": "...",
        "economic_impact": "...",
        "location": "...",
        "time": "...",
        "duration": "...",
        "confidence": "..."
    }]

    Each nested dictionary describes one failure case.
    DO NOT add commentary or text outside the JSON.

    Examples:
    ###
    Context: More than 130 km of motorways were closed directly after the event, of which 50 km were still closed two months later, with an estimated repair cost of EUR100 million (Hauser, 2021). Of the 112 bridges in the flooded 40 km of the Ahr valley (Rhineland-Palatinate), 62 bridges were destroyed, 13 were severely damaged and only 35 were in operation a month after the flood event (MDR, 2021).
    Question: How many bridges were destroyed in the Ahr valley during the 2021 flood event?
    Answer: 62
    ###
    Context: More than 130 km of motorways were closed directly after the event, of which 50 km were still closed two months later, with an estimated repair cost of EUR100 million (Hauser, 2021). Of the 112 bridges in the flooded 40 km of the Ahr valley (Rhineland-Palatinate), 62 bridges were destroyed, 13 were severely damaged and only 35 were in operation a month after the flood event (MDR, 2021).
    Question: How many bridges were in operation a month after the flood event in the Ahr valley?
    Answer: 35
    ###
    Context: More than 130 km of motorways were closed directly after the event, of which 50 km were still closed two months later, with an estimated repair cost of EUR100 million (Hauser, 2021). Of the 112 bridges in the flooded 40 km of the Ahr valley (Rhineland-Palatinate), 62 bridges were destroyed, 13 were severely damaged and only 35 were in operation a month after the flood event (MDR, 2021).
    Question: How many bridges were at least affected by the flood event in the Ahr valley?
    Answer: 77
    ###
    Context:  In total, at least 220 casualties have been reported, with insured loss estimates of approximately EUR 150 million–EUR 250 million in the Netherlands (Verbond voor Verzekeraars, 2022), EUR 2.2 billion in Belgium (Assuralia, 2022) and EUR 8.2 billion (GDV, 2022) in Germany. The event caused major damages to residential and commercial structures and to many critical infrastructure (CI) assets. 
    Question: How high are the estimated insured losses in Germany?
    Answer: EUR 8.2 billion
    ###
    Context: {context}
    Question: {question}
    Answer:

"""

template = Template(prompt_template)


# example_context = doc[6].page_content
# chunk_id = 6
# context = [
#     {
#         "text": example_context,
#         "citation": "Koks et al., 2022",
#         "ci_locations": df_ci_geo.loc[df_ci_geo["chunk_id"]==chunk_id],#to_dict(orient="records")
#     },  # TODO use author names or Primary keys from DB
#     # {"text": context, "citation": "Meier et al., 2025"},
# ]

# rendered_prompt = template.render(
#     context=context,
#     question=question,
#     # messages=messages
# )
# print(rendered_prompt)

In [ ]:

class DecoderModel:

    def __init__(self):

        login(token=os.environ["HUGGINGFACE_TOKEN"])  # TODO replace by using pydantic settings

        model_name = "EleutherAI/gpt-j-6B" #"distilbert-base-multilingual-cased"
        base_dir = "./huggingface_mirror"  # use default dir in .cache/
        model_dir = base_dir + "/hub/"  # + "models--" + model_name.replace("/", "--")
        print(model_dir)

        # quantization config
        # Load model with 4-bit quantization if applicable (use 4-bit integer instead of 32b floats) --> reduce the required VRAM for model application
        # see, https://huggingface.co/docs/transformers/quantization
        bnb_config = BitsAndBytesConfig(
            load_in_4bit=True,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_quant_type="nf4",
            bnb_4bit_compute_dtype=torch.float16,
        )

        self.pipeline, self.tokenizer = self.initialize_model(
            model_name, model_dir, bnb_config
        )

    def initialize_model(self, model_name: str, model_dir: str = None, bnb_config=None):

        # Model and Tokenizer initialization
        if not os.path.exists(model_dir):
            print("Model directory not found. Downloading model...")
            os.makedirs(model_dir, exist_ok=True)

            device = transformers.infer_device()
            print(f"Using device: {device}")
            model = GPTJForQuestionAnswering.from_pretrained(
                model_name,
                dtype="auto",
                attn_implementation="flash_attention_2",  # use with 4-bit quantization,
                # --> flash attention enables to use much larger sequence lengths without running into OOM issues
                quantization_config=bnb_config,
                # max_memory={0: "2GB", 1: "10GB"},  # distribute memory across GPUs
                # tp_plan="auto",
            )
            model.save_pretrained(model_dir)
            tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)
            tokenizer.save_pretrained(model_dir)

            print("Downloaded model and tokenizer")

        else:
            print(f"Using locally saved model from {model_dir}")

            model = GPTJForQuestionAnswering.from_pretrained(
                model_name,
                cache_dir=model_dir,
                local_files_only=True,  # tp_plan="auto" # set tensor parallel model (ie. splits model on multiple GPU)
                # dtype=torch.float16,
                attn_implementation="flash_attention_2",  # use with 4-bit quantization,
                # --> flash attention enables to use much larger sequence lengths without running into OOM issues
                quantization_config=bnb_config,
                max_memory={0: "2GB", 1: "10GB"},  # distribute memory across GPUs
                # tp_plan="auto",  # automatically use a tensor parallelism plan based on predefined configuration of the model (i.e. partition model on both GPUs)
            )
            print("Tensor parallel plan:", model._tp_plan)

            tokenizer = AutoTokenizer.from_pretrained(
                model_name, use_fast=True, cache_dir=model_dir, # use fast Rust-based tokenizer, when possible
            )

        # reduce further memory usage
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        model = model.to(device)
        model.use_checkpointing = True

        gc.collect()
        torch.cuda.empty_cache()

        # Pipeline setup for question answering
        pipeline = transformers.pipeline(  # load model locally from wsl .cache\
            "question-answering",  # task defining which pipeline is returned
            model=model,
            tokenizer=tokenizer, #(return_tensors="pt"),  # load specific tokenizer based on model-name (via AutoTokenizer) ensuring text is tokenized in accordance to the way the model was trained
            max_new_tokens=256,
            dtype=torch.float16,
            # low_cpu_mem_usage=True,
            device_map="auto",
        )
        return pipeline, tokenizer

    def generate_response(self, question: str, context: str, df_ci_geo: pd.DataFrame, chunk_id: int ):
        ## Preparing the input prompts

        context = [
            {
                "text": context,
                "citation": "Koks et al., 2022",
                "ci_locations": df_ci_geo.loc[
                    df_ci_geo["chunk_id"] == chunk_id
                ],  # to_dict(orient="records")
            },  # TODO use author names or Primary keys from DB
            # {"text": context, "citation": "Meier et al., 2025"},
        ]
        rendered_prompt = template.render(
            context=context,
            question=question,
        )
        print(f"Generating response for chunk_id: {chunk_id} ...")


        # Generating responses
        sequences = self.pipeline(
            question=question, context=rendered_prompt, # for GPT-J QA # TODO fix workaround with rendered rompt also incl Questoin besides insturctions
            max_new_tokens=256,
            do_sample=True,
            eos_token_id=self.tokenizer.eos_token_id,
        )
        # Extracting and returning the generated text
        return sequences


In [ ]:
## load doc
PARSED_TEXT_DIR = "../" + s.settings.PATH_DATA + "parsed_documents/"
FILE_PATH = PARSED_TEXT_DIR + "Koks et al 2022 Brief communication_cleaned.md"
loader = DoclingLoader(FILE_PATH)  # use chunks from Docling.Loader
doc = loader.load()

## init GPT model
decoder_model = DecoderModel()

# TODO make as pydantic object
df_responses = pd.DataFrame(
    columns=[
        "chunk_id",
        "infrastructure_type",
        "damage",
        # "societal_impact",
        "economic_impact",
        "location",
        "time",
        "duration",
        "confidence",
        "confidence_explanation",
        # "text_snippet",
    ]
)

## apply LLM on each chunk
## TODO replace iteration by loading entire document and use recursive chunking from langchain
for i, chunk in enumerate(doc):
    
    if df_ci_geo.loc[df_ci_geo["chunk_id"] == i].empty:
        continue

    context = [
        {
            "text": chunk.page_content,
            "citation": "Koks et al., 2022",
            "ci_locations": df_ci_geo.loc[df_ci_geo["chunk_id"] == i]
        },  # TODO use author names or Primary keys from DB
        # {"text": context, "citation": "Meier et al., 2025"},
    ]
    response_gpt = decoder_model.generate_response(
        question=question, context=context[0]["text"],
        df_ci_geo=df_ci_geo, chunk_id=i
    )

    resp = response_gpt["answer"]
    try:
        resp = (
            resp.split("]")[0] + "]"
        )  # remove potentially text outside json object and close the JSON array properlyt
        resp = pd.read_json(StringIO(resp))
        resp["chunk_id"] = i  # add chunk id as identifier
        df_responses = pd.concat([df_responses, resp], ignore_index=True)
    except ValueError as e:
        print(f"Cannot add response to output dataframe: {e}, \n{resp}")

2025-12-03 00:34:23,607 - INFO - Going to convert document batch...
2025-12-03 00:34:23,608 - INFO - Initializing pipeline for SimplePipeline with options hash 4cc01982ae99b46a2a63fcda46c47c35
2025-12-03 00:34:23,608 - INFO - Processing document Koks et al 2022 Brief communication_cleaned.md
2025-12-03 00:34:23,651 - INFO - Finished converting document Koks et al 2022 Brief communication_cleaned.md in 0.04 sec.
Token indices sequence length is longer than the specified maximum sequence length for this model (556 > 512). Running this sequence through the model will result in indexing errors


./huggingface_mirror/hub/
Using locally saved model from ./huggingface_mirror/hub/


`low_cpu_mem_usage` was None, now default to True since model is quantized.
Some weights of GPTJForQuestionAnswering were not initialized from the model checkpoint at EleutherAI/gpt-j-6B and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Tensor parallel plan: None


Device set to use cuda:0


Generating response for chunk_id: 1 ...


The input hidden states seems to be silently casted in float32, this might be related to the fact you have upcasted embedding or layer norm layers in float32. We will cast back the input in torch.float16.


Cannot add response to output dataframe: Expected object or value, 
 Answer]
Generating response for chunk_id: 3 ...
Cannot add response to output dataframe: Expected object or value, 


    The examples below should guide you in creating your answer]
Generating response for chunk_id: 5 ...
Cannot add response to output dataframe: Expected object or value, 
 Answer]
Generating response for chunk_id: 6 ...
Cannot add response to output dataframe: Expected object or value, 
 Answer]
Generating response for chunk_id: 7 ...
Cannot add response to output dataframe: Expected object or value, 


    The examples below should guide you in creating your answer]
Generating response for chunk_id: 8 ...
Cannot add response to output dataframe: Expected object or value, 


    The examples below should guide you in creating your answer]
Generating response for chunk_id: 9 ...
Cannot add response to output dataframe: Expected object or value, 


    The examples below should guide you in creating yo

You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset


Cannot add response to output dataframe: Expected object or value, 


    The examples below should guide you in creating your answer]
Generating response for chunk_id: 13 ...
Cannot add response to output dataframe: Expected object or value, 
 Answer]
Generating response for chunk_id: 14 ...
Cannot add response to output dataframe: Expected object or value, 
 Answer]
Generating response for chunk_id: 15 ...
Cannot add response to output dataframe: Expected object or value, 


    The examples below should guide you in creating your answer]
Generating response for chunk_id: 16 ...
Cannot add response to output dataframe: Expected object or value, 


    The examples below should guide you in creating your answer]
Generating response for chunk_id: 17 ...
Cannot add response to output dataframe: Expected object or value, 
 Answer]
Generating response for chunk_id: 18 ...
Cannot add response to output dataframe: Expected object or value, 


    The examples below should guide you in creat

In [ ]:
safety_response_gpt = df_responses.copy()

"system: [{'text': 'Nat. Hazards Earth Syst. Sci., 22, 3831–3838, 2022 https://doi.org/10.5194/nhess-22-3831-2022 © Author(s) 2022. This work is distributed under the Creative Commons Attribution 4.0 License.\\nBrief communication: Critical infrastructure impacts of the 2021 mid-July western European flood event\\nElco E. Koks1,2, Kees C. H. van Ginkel3,1, Margreet J. E. van Marle3, and Anne Lemnitzer4 1Institute for Environmental Studies, Vrije Universiteit Amsterdam, the Netherlands 2Oxford Programme for Sustainable Infrastructure Systems, Environmental Change Institute, University of Oxford, Oxford, United Kingdom 3Deltares, Delft, the Netherlands 4University of California, Irvine, Irvine, California, United States of America\\nCorrespondence: Kees C. H. van Ginkel (kees.vanginkel@deltares.nl)\\nReceived: 17 December 2021 – Discussion started: 23 December 2021 Revised: 10 August 2022 – Accepted: 18 October 2022 – Published: 29 November 2022', 'citation': 'Koks et al., 2022'}]\nuser: 

'Nat. Hazards Earth Syst. Sci., 22, 3831–3838, 2022 https://doi.org/10.5194/nhess-22-3831-2022 © Author(s) 2022. This work is distributed under the Creative Commons Attribution 4.0 License.\nBrief communication: Critical infrastructure impacts of the 2021 mid-July western European flood event\nElco E. Koks1,2, Kees C. H. van Ginkel3,1, Margreet J. E. van Marle3, and Anne Lemnitzer4 1Institute for Environmental Studies, Vrije Universiteit Amsterdam, the Netherlands 2Oxford Programme for Sustainable Infrastructure Systems, Environmental Change Institute, University of Oxford, Oxford, United Kingdom 3Deltares, Delft, the Netherlands 4University of California, Irvine, Irvine, California, United States of America\nCorrespondence: Kees C. H. van Ginkel (kees.vanginkel@deltares.nl)\nReceived: 17 December 2021 – Discussion started: 23 December 2021 Revised: 10 August 2022 – Accepted: 18 October 2022 – Published: 29 November 2022'

OutOfMemoryError: CUDA out of memory. Tried to allocate 128.00 MiB. GPU 0 has a total capacity of 5.67 GiB of which 121.75 MiB is free. Process 2232656 has 112.18 MiB memory in use. Including non-PyTorch memory, this process has 5.37 GiB memory in use. Of the allocated memory 5.13 GiB is allocated by PyTorch, and 125.64 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

## LLama

###  Prompt engineering

In [6]:

question = "Which impacts of infrastructure failures are mentioned in the text? Categorize the output by the type of infrastructure, societal or economic impacts, the location and possibly the time of the infrastructure failure."

## without s+e impacts
# question = "Which infrastructure failures are mentioned in the text? Categorize the output by the type of infrastructure, the location and possibly the time of the infrastructure failure."

In [7]:
# TODO move template to separate file and load via get_template(), define conditions (e.g. user is technical or not)
# TODo make pydantic class model for expected JSON output

# Example code: https://medium.com/@alecgg27895/jinja2-prompting-a-guide-on-using-jinja2-templates-for-prompt-management-in-genai-applications-e36e5c1243cf
# test instead of user_type (see: {% block user_type %}) the modification of question in regard to CI impact types (Tier 1,2,3 and 4 )


## first
#     You are an expert analyst assistant and should use ONLY the provided context to answer the following question:

## Text_snippet column
#     In the field "text_snippet" provide the exact linenumbers (within a list) from the context which you used to extract the information about the infrastructure failure and its impacts.
## too long responses


# prompt_template = """

#     You are an expert analyst assistant and should use ONLY the provided context to answer the following question:
    
#     Question: "{{ question }}"
       
#     Context:
#     {% for item in context %}
#     - {{ item.text }} (Citation: {{ item.citation }})
#     {% endfor %}

#     Evaluate and improve your answer based on the information about critical infrastructure (CI) types (column: "ci_entity") and their geolocations (column: "geo_entity") mentioned in CI locations.
    
#     CI locations:
#     {% for item in context %}
#     - ("ci_entity" and "geo_entity":\n {{ item["ci_locations"][["ci_entity", "geo_entity"]] }})
#     {% endfor %}


#     In the field "confidence" you should give an estimate how confident you are about the provided information on a scale between 1 (low) to 5 (high). 
#     In the "confidence_explanation" field give also a short explanation why you decided for a certain confidence value (maximum two bullet points within a list).
    

#     Return ONLY valid JSON in the following list format:
#     [{
#         "infrastructure_type": "...",
#         "damage": "...",
#         "location": "...",
#         "time": "...",
#         "duration": "...",
#         "confidence": "...",
#         "confidence_explanation": "[...]"
#     }]

#     Each nested dictionary describes one failure case.
#     DO NOT add commentary or text outside the JSON.
    

#     Answer:
# """

######################################################################


prompt_template = """

    You are an expert analyst assistant and should use ONLY the provided context to answer the following question:
    
    Question: "{{ question }}"
       
    Context:
    {% for item in context %}
    - {{ item.text }} (Citation: {{ item.citation }})
    {% endfor %}

    
    For the the fields "societal_impact" and "economic_impact" you should try to extract information about societal or economic consequences of infrastructure failures mentioned in the context.
    However, if you do not find any information about societal or economic consequences, then return for these fields a "NAN" value.
    
    For the fields ""infrastructure_type" and "location" you should evaluate and improve your answer based on the information mentioned in CI locations.

    CI locations:
    {% for item in context %}
    - ("ci_entity" and "geo_entity":\n {{ item["ci_locations"][["ci_entity", "geo_entity"]] }})
    {% endfor %}


    In the field "confidence" you should give an estimate how confident you are about the provided information in regard to the societal and economic impacts on a scale between 1 (low) to 5 (high). 
    In the "confidence_explanation" field give also a very short explanation why you decided for a certain confidence value (maximum two bullet points within a list).


    Return ONLY valid JSON in the following list format:
    [{
        "infrastructure_type": "...",
        "damage": "...",
        "societal_impact": "...",
        "economic_impact": "...",
        "location": "...",
        "confidence_explanation": "[...]",
        "confidence": "...",
    }]

    Each nested dictionary describes one failure case.
    DO NOT add commentary or text outside the JSON.
    

    Answer:
"""

template = Template(prompt_template)


# example_context = doc[6].page_content
# chunk_id = 6
# context = [
#     {
#         "text": example_context,
#         "citation": "Koks et al., 2022",
#         "ci_locations": df_ci_geo.loc[df_ci_geo["chunk_id"]==chunk_id],#to_dict(orient="records")
#     },  # TODO use author names or Primary keys from DB
#     # {"text": context, "citation": "Meier et al., 2025"},
# ]

# rendered_prompt = template.render(
#     context=context,
#     question=question,
#     # messages=messages
# )
# print(rendered_prompt)


## left overs
#  Try to be as specific as possible in your answer (bullet points), mention the impacts as numerical information along the location of the impact, and refer to the citations provided in the context.
# # Extract information about infrastructure failures based on the following question:

### Apply LLama on chunks 
* extract CI-GEO pairs (ie. most likely geolocation of each CI_TYPE )
* pass NER table to prompt for evaluating and improving LLM response
* Test approach by applying it on three cleaned documents


In [8]:
# # empty CUDA cache
import gc
import torch

gc.collect()

torch.cuda.empty_cache()
# print(torch.cuda.memory_summary(device=None, abbreviated=False))

In [9]:
# init class for decoder and tokenizer


class DecoderModel:
    def __init__(self):
        login(
            token=os.environ["HUGGINGFACE_TOKEN"]
        )  # TODO replace by using pydantic settings

        # model_name = "google/gemma-3-4b-it" # "kallidavidson/TinyBERT_General_4L_312D"  # "huawei-noah/TinyBERT_General_4L_312D" # - for QA - less DWL
        model_name = "meta-llama/Llama-2-7b-chat-hf"
        # model_name = "EleutherAI/gpt-j-6B" #"distilbert-base-multilingual-cased"
        base_dir = "./huggingface_mirror"  # use default dir in .cache/
        model_dir = base_dir + "/hub/"  # + "models--" + model_name.replace("/", "--")
        print(model_dir)

        # quantization config
        # Load model with 4-bit quantization if applicable (use 4-bit integer instead of 32b floats) --> reduce the required VRAM for model application
        # see, https://huggingface.co/docs/transformers/quantization
        bnb_config = BitsAndBytesConfig(
            load_in_4bit=True,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_quant_type="nf4",
            bnb_4bit_compute_dtype=torch.float16,
        )

        self.pipeline, self.tokenizer = self.initialize_model(
            model_name, model_dir, bnb_config
        )

    def initialize_model(self, model_name: str, model_dir: str = None, bnb_config=None):
        # Model and Tokenizer initialization
        if not os.path.exists(model_dir):
            print("Model directory not found. Downloading model...")
            os.makedirs(model_dir, exist_ok=True)

            device = transformers.infer_device()
            print(f"Using device: {device}")
            model = AutoModelForCausalLM.from_pretrained(
                model_name,
                dtype="auto",
                attn_implementation="flash_attention_2",  # use with 4-bit quantization,
                # --> flash attention enables to use much larger sequence lengths without running into OOM issues
                quantization_config=bnb_config,
                # max_memory={0: "2GB", 1: "10GB"},  # distribute memory across GPUs
            )
            model.save_pretrained(model_dir)
            tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)
            tokenizer.save_pretrained(model_dir)

            print("Downloaded model and tokenizer")

        else:
            print(f"Using locally saved model from {model_dir}")

            model = AutoModelForCausalLM.from_pretrained(
                model_name,
                cache_dir=model_dir,
                local_files_only=True,  # tp_plan="auto" # set tensor parallel model (ie. splits model on multiple GPU)
                # dtype="auto",
                attn_implementation="flash_attention_2",  # use with 4-bit quantization,
                # --> flash attention enables to use much larger sequence lengths without running into OOM issues
                quantization_config=bnb_config,
                # tp_plan="auto",  # automatically use a tensor parallelism plan based on predefined configuration of the model (i.e. partition model on both GPUs)
            )
            # print("Tensor parallel plan:", model._tp_plan)

            tokenizer = AutoTokenizer.from_pretrained(
                model_name,
                use_fast=True,
                cache_dir=model_dir,  # use fast Rust-based tokenizer, when possible
            )

        # reduce further memory usage
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        model = model.to(device)
        model.use_checkpointing = True

        torch.cuda.empty_cache()

        # Pipeline setup for question answering
        pipeline = transformers.pipeline(  # load model locally from wsl .cache\
            "text-generation",
            # "question-answering",  # task defining which pipeline is returned
            model=model,
            tokenizer=tokenizer,
            # (return_tensors="pt"),  # load specific tokenizer based on model-name (via AutoTokenizer) ensuring text is tokenized in accordance to the way the model was trained
            max_new_tokens=1024, # high max toke otherwise output is truncated
            device_map="auto",
        )
        return pipeline, tokenizer

    def generate_response(
        self, question: str, context: list, df_ci_geo: pd.DataFrame, chunk_id: int
    ):
    
        rendered_prompt = template.render(
            context=context,
            question=question,
        )
        print(f"Generating response for chunk_id: {chunk_id} ...")

        sequences = self.pipeline(
            rendered_prompt,  # jinja template
            max_new_tokens=1024, # use default to not truncate the LLM response
            do_sample=True,
            num_beams=1,  # select token based on probability distribution over entire model’s vocabulary
            # top_k=10,
            # top_p=0.5,
            temperature=0.2,
            # num_return_sequences=1,
            eos_token_id=self.tokenizer.eos_token_id,
            return_full_text=False,  # allow bullet point answers
        )
        # Extracting and returning the generated text
        return sequences

In [ ]:

PARSED_TEXT_DIR = "../" + s.settings.PATH_DATA + "parsed_documents/"


## call language model for recognition of CI and geolocation
## see for more info: https://spacy.io/usage/rule-based-matching#entityruler
## NOTE EntityRuler is hidden inside .add_pipe()
# NOTE Creating new entity (CI_TYPE) solves the issue that FAC entities (buildings, airports, highways, bridges, etc.) refer only to the name of the facility (e.g. A76, Ahrtalbahn)
config = {"spans_key": None, "annotate_ents": True, "overwrite": False}
try:
    ruler = nlp.add_pipe("span_ruler", config=config)
    ruler.from_disk("../ner_patterns.jsonl")
except ValueError:
    print("SpanRuler already exists in pipeline.")
    ruler = nlp.get_pipe("span_ruler")
    ruler.from_disk("../ner_patterns.jsonl")


df_responses = pd.DataFrame(
    columns=[
        "chunk_id",
        "infrastructure_type",
        "damage",
        "societal_impact",
        "economic_impact",
        "location",
        # "time",
        # "duration",
        "confidence",
        "confidence_explanation",
        # "text_snippet",
    ]
)

## init LLM pipeline
decoder_model = DecoderModel()



## iterate over documents
for i, filename in enumerate(glob.glob(str(Path(PARSED_TEXT_DIR, "*cleaned.md")))):

    no_documents = len(glob.glob(str(Path(PARSED_TEXT_DIR, "*cleaned.md"))))
    filepath = Path(filename)
    filename_stem = filepath.stem


    print(f"\n\n ######## -------- Processing document [{i+1}/{no_documents}]: {filepath.name} -------- ######## \n")

    ## extract authors, publication year and title 
    citation_pattern = r"(.*?)(\d{4})(.*)" # split at first occurrence of year
    try:
        authors, year, title = re.findall(citation_pattern, filename_stem)[0]
        citation = f"{authors} {year}"
    except AttributeError as e:
        print(f"Could not extract citation from title: {e}")
        citation = filename_stem


    print(f"\n ######## -------- Getting geolocation of CI assets -------- ######## \n")
    ## Create New entity for transport infrastructure and apply it on any doc

    ## TODO make as pydantic class with fixed attributes
    df_ci_geo = pd.DataFrame(
        columns=[
            "chunk_id",
            "ci_entity",
            "ci_entity_label",
            "geo_entity",
            "geo_entity_label",
            "token_distance",
        ]
    )

    ## load doc
    loader = DoclingLoader(filepath)  # use chunks from Docling.Loader
    doc = loader.load()
    doc = doc[5:15]

    ## get most likely geolocation for each CI entity based on distance
    for i, chunk in enumerate(doc):
        nlp_chunk = nlp(chunk.page_content)
        all_ents = [ent for ent in nlp_chunk.ents]
        ci_type_ents = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE", "FAC"]]
        ci_type_ents_info = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE"]]
        fac_ents_info = [ent for ent in nlp_chunk.ents if ent.label_ in ["FAC"]]

        # check if chunk contains CI_TYPE entities
        if len(ci_type_ents) > 0:
            print(f"\nChunk [{i}], No. CI_TYPE and FAC entities: {len(ci_type_ents)}")
            # print(
            #     f"Contains following entities for CI_TYPE: {ci_type_ents_info}, FAC: {fac_ents_info}"
            # )
            # print(f"Chunk text [{i}]:", chunk.page_content)

            # iterate over all entities within chunk
            for ent_idx in range(len(all_ents)):
                # when entity is CI_TYPE or FAC (i.e. buidling, airports, highways) do following ...
                if all_ents[ent_idx].label_ in ["CI_TYPE", "FAC"]:
                    ci_idx = ent_idx

                    ## .. calculate distances between CI_TYPE entity and  all GEO entities in chunk based on index position
                    distance_list = []
                    idx_in_chunk = []
                    try:
                        for ent_idx in range(len(all_ents)):
                            # TODO calc distances between CI_TYPE ~ GEO entities based on word numbers and not entities
                            if all_ents[ent_idx].label_ in ["GPE", "LOC"]:
                                geo_idx = ent_idx
                                dist_ent_pair = np.abs(ci_idx - geo_idx)
                                distance_list.append(dist_ent_pair)
                                idx_in_chunk.append((ent_idx))
                                closest_pair_idx = np.argmin(
                                    distance_list
                                )  # idx of closest GEO entity
                                distance_closest_pair = distance_list[closest_pair_idx]

                        threshold = 5  # max token distance between CI_TYPE and GEO entity
                        if distance_closest_pair > threshold:
                            print(
                                f""" Token distance between CI_TYPE/FAR and next GEO entity is {distance_closest_pair} and thus larger than the allowed distance of {threshold} tokens """
                            )
                            continue
                        else:
                            pass
                            # print(
                            #     f"""  Closest GEO entity to CI_TYPE/FAC entity "{all_ents[ci_idx]}" is "{all_ents[idx_in_chunk[closest_pair_idx]]}" at distance {distance_closest_pair}"""
                            # )

                        ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
                        result_dict = {
                            "chunk_id": i,
                            "ci_entity": all_ents[ci_idx].text,
                            "ci_entity_label": all_ents[ci_idx].label_,
                            "geo_entity": all_ents[idx_in_chunk[closest_pair_idx]].text,
                            "geo_entity_label": all_ents[
                                idx_in_chunk[closest_pair_idx]
                            ].label_,
                            "token_distance": distance_closest_pair,
                        }
                        df_ci_geo = pd.concat(
                            [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
                        )

                    except IndexError:
                        # print("No GEO entities found in this chunk.")
                        continue
                    # print("\nidx_in_chunk, closest pair idx", idx_in_chunk, closest_pair_idx)

                    # spacy.displacy.render(
                    #     nlp_chunk, style="ent",
                    #     options={"ents": ["CI_TYPE", "GPE", "LOC", "FAC"], "colors": {"CI_TYPE": "violet"}}
                    # )

            else:
                print("\nNo CI_TYPE or FAC entities found in this chunk.")
                continue


    print(f"\n  #############  -------- Text-2-Data: {filepath.name} -------- #############  \n")

    ## apply decoder on each chunk
    ## TODO replace iteration by loading entire document and use recursive chunking from langchain
    for j, chunk in enumerate(doc):

        if df_ci_geo.loc[df_ci_geo["chunk_id"] == j].empty:
            continue

        context = [
            {
                "text": chunk.page_content,
                "citation": citation,
                "ci_locations": df_ci_geo.loc[df_ci_geo["chunk_id"] == j],
            },  # TODO use author names or Primary keys from DB later
        ]

        response = decoder_model.generate_response(
            question=question, context=context, df_ci_geo=df_ci_geo, chunk_id=j
        )

        ## postprocess response
        resp = response[0]["generated_text"].replace("\n", "")
        try:
            resp = (resp.split("]")[0] + "]")  # remove potential text outside of json object
            df_resp = pd.read_json(StringIO(resp))
            df_resp["chunk_id"] = j  # add chunk id as identifier
            df_resp["citation"] = citation  # add citation info
            df_responses = pd.concat([df_responses, df_resp], ignore_index=True)
        except ValueError as e:
            print(f"Cannot add response to output dataframe: {e}, \n{resp}")



    # clean up after each document
    gc.collect()
    torch.cuda.empty_cache()

./huggingface_mirror/hub/
Using locally saved model from ./huggingface_mirror/hub/


`low_cpu_mem_usage` was None, now default to True since model is quantized.
Loading checkpoint shards: 100%|██████████| 2/2 [00:07<00:00,  3.55s/it]
Device set to use cuda:0




 ######## -------- Processing document [1/3]: Mohr 2022 A multi-disciplinary analysis of the exceptional flood event of July 2021 in central Europe - Part 1 Event desciption and analysis_cleaned.md -------- ######## 


 ######## -------- Getting geolocation of CI assets : Mohr 2022 A multi-disciplinary analysis of the exceptional flood event of July 2021 in central Europe - Part 1 Event desciption and analysis_cleaned.md -------- ######## 



2025-12-04 21:49:42,675 - INFO - Going to convert document batch...
2025-12-04 21:49:42,675 - INFO - Initializing pipeline for SimplePipeline with options hash 4cc01982ae99b46a2a63fcda46c47c35
2025-12-04 21:49:42,676 - INFO - Processing document Mohr 2022 A multi-disciplinary analysis of the exceptional flood event of July 2021 in central Europe - Part 1 Event desciption and analysis_cleaned.md
2025-12-04 21:49:42,860 - INFO - Finished converting document Mohr 2022 A multi-disciplinary analysis of the exceptional flood event of July 2021 in central Europe - Part 1 Event desciption and analysis_cleaned.md in 0.19 sec.
Token indices sequence length is longer than the specified maximum sequence length for this model (609 > 512). Running this sequence through the model will result in indexing errors



Chunk [0], No. CI_TYPE and FAC entities: 3
  Closest GEO entity to CI_TYPE/FAC entity "railways" is "Rhine" at distance 2
  Closest GEO entity to CI_TYPE/FAC entity "roads" is "Rhine" at distance 3
  Closest GEO entity to CI_TYPE/FAC entity "bridges" is "Rhine" at distance 4

No CI_TYPE or FAC entities found in this chunk.

Chunk [6], No. CI_TYPE and FAC entities: 2
No GEO entities found in this chunk.
No GEO entities found in this chunk.

No CI_TYPE or FAC entities found in this chunk.

  #############  -------- Text-2-Data: Mohr 2022 A multi-disciplinary analysis of the exceptional flood event of July 2021 in central Europe - Part 1 Event desciption and analysis_cleaned.md -------- #############  

Generating response for chunk_id: 0 ...


 ######## -------- Processing document [2/3]: Koks et al 2022 Brief communication_cleaned.md -------- ######## 


 ######## -------- Getting geolocation of CI assets : Koks et al 2022 Brief communication_cleaned.md -------- ######## 



2025-12-04 21:50:14,131 - INFO - Going to convert document batch...
2025-12-04 21:50:14,132 - INFO - Initializing pipeline for SimplePipeline with options hash 4cc01982ae99b46a2a63fcda46c47c35
2025-12-04 21:50:14,132 - INFO - Processing document Koks et al 2022 Brief communication_cleaned.md
2025-12-04 21:50:14,214 - INFO - Finished converting document Koks et al 2022 Brief communication_cleaned.md in 0.08 sec.
Token indices sequence length is longer than the specified maximum sequence length for this model (556 > 512). Running this sequence through the model will result in indexing errors



Chunk [0], No. CI_TYPE and FAC entities: 10
  Closest GEO entity to CI_TYPE/FAC entity "road" is "Germany" at distance 1
  Closest GEO entity to CI_TYPE/FAC entity "railway" is "Germany" at distance 2
  Closest GEO entity to CI_TYPE/FAC entity "motorways" is "Hauser" at distance 4
  Closest GEO entity to CI_TYPE/FAC entity "bridges" is "the Ahr valley" at distance 2
  Closest GEO entity to CI_TYPE/FAC entity "bridges" is "Rhineland-Palatinate" at distance 2
  Closest GEO entity to CI_TYPE/FAC entity "roads" is "the Ahr valley" at distance 2
  Closest GEO entity to CI_TYPE/FAC entity "bridges" is "the Ahr valley" at distance 1
  Closest GEO entity to CI_TYPE/FAC entity "A1" is "the Ahr valley" at distance 4
  Closest GEO entity to CI_TYPE/FAC entity "motorway" is "the Ahr valley" at distance 5
 Token distance between CI_TYPE/FAR and next GEO entity is 14 and thus larger than the allowed distance of 5 tokens 

No CI_TYPE or FAC entities found in this chunk.

Chunk [1], No. CI_TYPE and F

You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset


Generating response for chunk_id: 9 ...


 ######## -------- Processing document [3/3]: Korzilius 2021 Nach der Flut_cleaned.md -------- ######## 


 ######## -------- Getting geolocation of CI assets : Korzilius 2021 Nach der Flut_cleaned.md -------- ######## 



2025-12-04 21:56:36,567 - INFO - Going to convert document batch...
2025-12-04 21:56:36,568 - INFO - Initializing pipeline for SimplePipeline with options hash 4cc01982ae99b46a2a63fcda46c47c35
2025-12-04 21:56:36,568 - INFO - Processing document Korzilius 2021 Nach der Flut_cleaned.md
2025-12-04 21:56:36,601 - INFO - Finished converting document Korzilius 2021 Nach der Flut_cleaned.md in 0.04 sec.
Token indices sequence length is longer than the specified maximum sequence length for this model (513 > 512). Running this sequence through the model will result in indexing errors



Chunk [0], No. CI_TYPE and FAC entities: 1
No GEO entities found in this chunk.

No CI_TYPE or FAC entities found in this chunk.

Chunk [9], No. CI_TYPE and FAC entities: 1
  Closest GEO entity to CI_TYPE/FAC entity "Hospital" is "Erftstadt" at distance 1

No CI_TYPE or FAC entities found in this chunk.

  #############  -------- Text-2-Data: Korzilius 2021 Nach der Flut_cleaned.md -------- #############  

Generating response for chunk_id: 9 ...
Cannot add response to output dataframe: Unmatched ''"' when when decoding 'string', 
    [        {            "infrastructure_type": "Hospital",            "damage": "Zerstörung",            "societal_impact": "NAN",            "economic_impact": "NAN",            "location": "Erftstadt",            "confidence_explanation": "[The text does not provide any information about societal or economic consequences of the hospital failure.]


In [13]:
r = response[0]["generated_text"].replace("\n", "")

pd.read_json(StringIO(r))#.to_list()


,infrastructure_type,damage,societal_impact,economic_impact,location,confidence_explanation,confidence
0,bridges,3 bridges collapsed,disrupted transportation and communication,loss of business and revenue,Belgium,"Based on the information provided, it is likely that the failure of 3 bridges in Belgium had a significant impact on the transportation and communication networks in the area. The loss of business and revenue is also a likely consequence of the failure, as the bridges were likely used by important transportation routes and may have been relied upon by local businesses.",4
1,schools,2 schools damaged,disrupted education for students,loss of productivity and revenue for schools,Belgium,"Based on the information provided, it is likely that the damage to 2 schools in Belgium had a significant impact on the education of students in the area. The loss of productivity and revenue for schools is also a likely consequence of the damage, as schools may have been forced to close or operate at reduced capacity.",4
2,hospitals,1 hospital damaged,disrupted medical care for patients,loss of revenue for the hospital,Belgium,"Based on the information provided, it is likely that the damage to 1 hospital in Belgium had a significant impact on the medical care provided to patients in the area. The loss of revenue for the hospital is also a likely consequence of the damage, as hospitals may have been forced to reduce their operations or close temporarily.",4


#### response

In [12]:
safety_df = df_responses.copy()

# save to disk along with prompt text


OUTPUT_DIR = "../" + s.settings.PATH_DATA + "llm_outputs/"
os.makedirs(OUTPUT_DIR, exist_ok=True)

outfile_name = "ci_failure_impacts_responses_llama2_with_eco_explicit_3.csv"
outfile_path = OUTPUT_DIR +  outfile_name


if not os.path.isfile(outfile_path):
    outfile_response_path = Path(outfile_path)
    outfile_prompt_path = Path(OUTPUT_DIR +  "prompt_" + outfile_name.replace(".csv", ".txt"))

    print(f"Saving prompt and LLm response to {outfile_response_path.parent} ...")
    with open(outfile_prompt_path, "w") as f:
        f.write(prompt_template)
    safety_df.to_csv(outfile_response_path, index=False)
else:
    print(f"Output file {Path(outfile_name).stem} already exists. Skip saving to avoid overwriting ...")



Saving prompt and LLm response to ../../data/llm_outputs ...


In [13]:
## 10 min


df_ci_geo

,chunk_id,ci_entity,ci_entity_label,geo_entity,geo_entity_label,token_distance
0,9,Hospital,CI_TYPE,Erftstadt,GPE,1


## Evaluation

### Manual comparison CI_location_table vs LLm response


#### chunk 5
In Germany, road and railway infrastructure was severely damaged as documented exemplarily in Fig. 1. Cost estimates reach up to EURO 2 billion Euro (MDR, 2021). More than 130 km of motorways were closed directly after the event, of which 50 km were still closed two months later, with an estimated repair cost of EUR 100 million (Hauser, 2021). Of the 112 bridges in the flooded 40 km of the Ahr valley (Rhineland-Palatinate), 62 bridges were destroyed, 13 were severely damaged and only 35 were in operation a month after the flood event (MDR, 2021). Over 74 km of roads, paths and bridges in the Ahr valley have been (critically) damaged. In some cases, repairs are expected to take months to years (Zeit Online, 2021). For example, major freeway sections, including parts of the A1 motorway, were closed until early 2022 (24Rhein, 2022). In addition, about 50 000 cars were damaged, causing insurance claims of some EUR 450 million (ADAC, 2021). The German railway provider Deutsche Bahn expects asset damages of around EUR 1.3 billion. Among other things, 180 level crossings, almost 40 signal'

In [ ]:
doc[5].page_content

In [ ]:
df_ci_geo[df_ci_geo["chunk_id"] == 5]

In [ ]:
df_responses[df_responses["chunk_id"] == 5]

In [ ]:
print(
    f"CI_TYPE \n LLM responses:\n {df_responses.infrastructure_type.unique()}, \n\n NER pairs:\n {df_ci_geo.ci_entity.unique()}"
)

In [ ]:
print(
    f"LOCATION:\n LLm responses:\n {df_responses.location.unique()}, \n\n NER pairs:\n {df_ci_geo.geo_entity.unique()}"
)

### chunk 13 +14
'We found no information regarding direct impact on solid-waste facilities as a result of the flood event. However, there is a large pressure on the solid-waste sector to clean the affected areas; 1 month after the event, we observed dozens of large temporary waste fills and frequent incidences of oil pollution in Rhineland-Palatinate during a field visit. In the Ahrweiler district alone, the flood caused as much solid waste as normally would be collected over 30 years. In Belgium, the amount of solid waste is estimated around 160 000 t, stored at several places, such as the abandoned highway track A601. This highway has been used for approximately 9 months as a temporary storage for debris (Couplez, 2022). In the Netherlands, there have been primarily problems with waste deposits along the river banks, which is mostly the solid waste transported by the river from further upstream. Thousands of tonnes of tree debris (logs and\nNat. Hazards Earth Syst. Sci., 22, 3831–3838, 2022\nE. E. Koks et al.: Flood impacts to infrastructure'


'of running water and electricity (Ärzte Zeitung, 2021). After 1.5 months, medical care was guaranteed again in the most affected regions in Rhineland-Palatinate (Hochwasser Ahr, 2021c). In the state of North Rhine-Westphalia, approximately 68 hospitals have been affected, of which several have been affected severely and will take at least 1.5 years to be rebuilt (Fig. 2). Direct damages are estimated to be at least EUR 100 million to repair all medical facilities (Korzilius, 2021). In the town of Eschweiler (Germany), for example, the basement of the hospital was flooded, as well as the outbuildings and the entire outdoor area. The power supply collapsed, the entire building technology was destroyed and some 300 patients had to be evacuated by helicopter. Property damage is expected to be around EUR 50 million. Within 3.5 weeks, the hospital was partly operational, and within 3 months, all hospital operations continued normally (SAH Eschweiler, 2021). The Mutterhaus Ehrang hospital in Trier (Germany) is now permanently closed as the hospital is too severely damaged to rebuild. Furthermore, in the region of Rhineland-Palatinate (Germany), 19'


In [ ]:
doc[14].page_content

'of running water and electricity (Ärzte Zeitung, 2021). After 1.5 months, medical care was guaranteed again in the most affected regions in Rhineland-Palatinate (Hochwasser Ahr, 2021c). In the state of North Rhine-Westphalia, approximately 68 hospitals have been affected, of which several have been affected severely and will take at least 1.5 years to be rebuilt (Fig. 2). Direct damages are estimated to be at least EUR 100 million to repair all medical facilities (Korzilius, 2021). In the town of Eschweiler (Germany), for example, the basement of the hospital was flooded, as well as the outbuildings and the entire outdoor area. The power supply collapsed, the entire building technology was destroyed and some 300 patients had to be evacuated by helicopter. Property damage is expected to be around EUR 50 million. Within 3.5 weeks, the hospital was partly operational, and within 3 months, all hospital operations continued normally (SAH Eschweiler, 2021). The Mutterhaus Ehrang hospital in 

In [ ]:
df_ci_geo[df_ci_geo["chunk_id"].isin([13, 14])]

,chunk_id,ci_entity,ci_entity_label,geo_entity,geo_entity_label,token_distance
53,13,waste facilities,CI_TYPE,Rhineland-Palatinate,GPE,3
54,13,solid waste,CI_TYPE,Ahrweiler,LOC,1
55,13,solid waste,CI_TYPE,Belgium,GPE,1
56,13,highway,CI_TYPE,Belgium,GPE,3
57,13,A601,FAC,Belgium,GPE,4
58,13,highway,CI_TYPE,Netherlands,GPE,4
59,13,solid waste,CI_TYPE,Netherlands,GPE,1
60,14,hospitals,CI_TYPE,North Rhine-Westphalia,GPE,2
61,14,hospital,CI_TYPE,Germany,GPE,1
62,14,hospital,CI_TYPE,Germany,GPE,5


In [ ]:
df_responses[df_responses["chunk_id"].isin([13, 14])]

,chunk_id,infrastructure_type,damage,economic_impact,location,time,duration
31,13,waste facilities,dozens of large temporary waste fills and frequent incidences of oil pollution,as much solid waste as normally would be collected over 30 years,Rhineland-Palatinate,1 month after the event,approximately 30 years
32,13,highway,used for approximately 9 months as a temporary storage for debris,thousands of tonnes of tree debris,Belgium,9 months after the event,approximately 30 years
33,13,solid waste,caused as much solid waste as normally would be collected over 30 years,estimated around 160 000 t,Ahrweiler district,1 month after the event,approximately 30 years
34,13,river banks,primarily problems with waste deposits,thousands of tonnes of tree debris,Netherlands,after the event,approximately 30 years
35,14,hospitals,"approximately 68 hospitals have been affected, of which several have been affected severely and will take at least 1.5 years to be rebuilt",direct damages are estimated to be at least EUR 100 million to repair all medical facilities,North Rhine-Westphalia,after 1.5 months,at least 1.5 years
36,14,hospitals,"the basement of the hospital was flooded, as well as the outbuildings and the entire outdoor area",property damage is expected to be around EUR 50 million,Eschweiler (Germany),within 3.5 weeks,within 3 months
37,14,hospital,The Mutterhaus Ehrang hospital in Trier (Germany) is now permanently closed as the hospital is too severely damaged to rebuild,"furthermore, in the region of Rhineland-Palatinate (Germany), 19'",Trier,now,


In [ ]:
print(
    f"CI_TYPE \n LLM responses:\n {df_responses.infrastructure_type.unique()}, \n\n NER pairs:\n {df_ci_geo.ci_entity.unique()}"
)

In [ ]:
print(
    f"LOCATION:\n LLm responses:\n {df_responses.location.unique()}, \n\n NER pairs:\n {df_ci_geo.geo_entity.unique()}"
)

## Test alternative approaches for entity linking / relation extraction

In [ ]:
import torch
import gc

print(torch.cuda.memory_summary(device=None, abbreviated=False))
# # empyty CUDA cache
gc.collect()

torch.cuda.empty_cache()
# print(torch.cuda.memory_summary(device=None, abbreviated=False))